### PR TITLE
[RFC] Pull request for  ip_k66f board 13.04

### PR DESCRIPTION
**Pull request by lmajewski**
Dear Community,

This PR is more a RFC for proposed changes for NXP's Kinetics MCUX ENET to add support for
KSZ8794 DSA switch.

Most notable changes:
- Fixed clock timing - sleep now works correctly on ip_k66f
- Restructure the Kconfig.defconfig and ip_k66f_defconfig
- The MII interrupt has been disabled. It is not needed as we don't use SMI for PHY communication and we solely use SPI.
- Support for ETH MCUX fixed PHY (Similar to Linux kernel)
- Introduction of DSA infrastructure to Zephyr
- KSZ8794 setup and configuration

I would like to hear your opinion on the FIXED_PHY and proposed DSA support.

### CODE REVIEW

File: subsys/net/ip/ipv4_autoconf.c [hunk 1]
```diff
@@ -168,12 +168,12 @@ static void ipv4_autoconf_send(struct net_if_ipv4_autoconf *ipv4auto)
...
-		ipv4auto->requested_ip.s4_addr[0] = 169U;
-		ipv4auto->requested_ip.s4_addr[1] = 254U;
-		ipv4auto->requested_ip.s4_addr[2] = sys_rand32_get() % 254;
+		ipv4auto->requested_ip.s4_addr[0] = 192U;
+		ipv4auto->requested_ip.s4_addr[1] = 168U;
+		ipv4auto->requested_ip.s4_addr[2] = 0U;
 		ipv4auto->requested_ip.s4_addr[3] = sys_rand32_get() % 254;
 
-		NET_DBG("%s: Starting probe for 169.254.%d.%d", "Init",
+		NET_DBG("%s: Starting probe for 192.168.%d.%d", "Init",
```
Comment (jukkar) [inline subsys/net/ip/ipv4_autoconf.c:176]: Nack to this change. The IPv4 auto addresses are from 169.254/16 subnet. See https://tools.ietf.org/html/rfc3927 for details.

**Review by jukkar — CHANGES_REQUESTED** [review]
No objections to adding DSA support to Zephyr. There is actually an issue #22061 for this so you could mention that issue in one of the commits and add "Fixes ..." to it.
In order to help reviewing this thing, I would split the PR into multiple pieces:
* the DSA infrastructure support
* changes to boards
* changes to mcux

Currently many of the commits are somewhat unrelated to each other so in order to ease the review, it would be great if you could split the PR into more managable pieces.

File: subsys/net/ip/ipv4_autoconf.c (hunk 1 above)
Comment (lmajewski) [reply subsys/net/ip/ipv4_autoconf.c:176]: Thanks for the feedback. Indeed I've used this change for testing.

Comment (lmajewski) [reply subsys/net/ip/ipv4_autoconf.c:176]: I will split this patch this set and fix issues indicated by gitlint, sanitycheck and uncrustify.

**Comment by AntoineZen** [comment]
Hi @lmajewski,

I'm about going the same route but with KSZ9893. In our setup, the MCU (i.MX-RT 1062) has two ethernet (MAC) ports, where one port is  having a single PHY (KSZ8081) and the second port is connected to the KSZ9893 trought RMII. The management of the switch is done over I2C and SMI (MDIO) is not used.  With you changes to `eth_mcux.c`, the configuration the where to use SMI or not is static, which enforce to use the same kind of management interface on both Ethernet port. This would be a problem with etherogenous management interfaces. Do you think it worth to make this configuration option a run-time/boot from the device-tree configuration ?

**Comment by lmajewski** [comment]
> Hi @lmajewski,
> 
> I'm about going the same route but with KSZ9893. In our setup, the MCU (i.MX-RT 1062) has two ethernet (MAC) ports, where one port is having a single PHY (KSZ8081) and the second port is connected to the KSZ9893 trought RMII. The management of the switch is done over I2C and SMI (MDIO) is not used. With you changes to `eth_mcux.c`, the configuration the where to use SMI or not is static (`CONFIG_ETH_MCUX_NO_PHY_SMI`), which enforce to use the same kind of management interface on both Ethernet port. This would be a problem with heterogeneous management interfaces. Do you think it worth to make this configuration option a run-time/boot from the device-tree configuration ?

In my use case (the single eth port on K66F) the static (in Kconfig) option was the most appropriate and straightforward.
However, as you described, you do have a different use case where both SMI and I2C are required.
That would be a good rationale to convert this option to be configurable via device tree, so each ETH port could be configured.

Note 1:
Just informative - I'm going to sent soon patches providing DSA infrastructure for Zephyr - I'm using KSZ8794 IC.

Note 2:
For your use case you can also use "fixed-phy" functionality recently merged:
https://github.com/zephyrproject-rtos/zephyr/pull/24460

**Pull request closed without merging**